src/Solvers/multigrid_solver.jl

```diff
@@ -128,30 +163,84 @@ function fill_matrix_elements!(A, template_field, linear_operator!, args...)
     return nothing
 end
 
+function fill_matrix_elements2!(template_field, linear_operator!, args...)
+    Nx, Ny, Nz = size(template_field)
+    make_column(f) = reshape(interior(f), Nx*Ny*Nz)
+
+    eᵢⱼₖ = similar(template_field)
+    ∇²eᵢⱼₖ = similar(template_field)
+
+    arch = architecture(template_field)
+    colptr = array_type(arch){Int64}(undef, Nx*Ny*Nz+1)
+    rowval = array_type(arch){Int64}(undef, 0)
+    nzval  = array_type(arch){Float64}(undef, 0)
+
+    CUDA.@allowscalar colptr[1] = 1
+
+    for k = 1:Nz, j in 1:Ny, i in 1:Nx
+        eᵢⱼₖ .= 0
+        ∇²eᵢⱼₖ .= 0
+        CUDA.@allowscalar eᵢⱼₖ[i, j, k] = 1
+        fill_halo_regions!(eᵢⱼₖ)
+        linear_operator!(∇²eᵢⱼₖ, eᵢⱼₖ, args...)
+        count = 0
+        for n = 1:Nz, m in 1:Ny, l in 1:Nx
+            # CUDA.@allowscalar value = ∇²eᵢⱼₖ[l, m, n]
+            CUDA.@allowscalar if ∇²eᵢⱼₖ[l, m, n] != 0
+                append!(rowval, Ny*Nx*(n-1) + Nx*(m-1) + l)
+                CUDA.@allowscalar append!(nzval, ∇²eᵢⱼₖ[l, m, n])
+                count += 1
+            end
+        end
+        CUDA.@allowscalar colptr[Ny*Nx*(k-1) + Nx*(j-1) + i + 1] = colptr[Ny*Nx*(k-1) + Nx*(j-1) + i] + count
+    end
+    if arch == CPU()
+        return Nx*Ny*Nz, Nx*Ny*Nz, colptr, rowval, nzval
+    end
+    return colptr, rowval, nzval, (Nx*Ny*Nz, Nx*Ny*Nz)
+end
+
 """
     solve!(x, solver::MultigridSolver, b; kwargs...)
 
 Solve `A * x = b` using a multigrid method, where `A * x` is determined
-by `solver.linear_operation`.
+by `solver.matrix`.
 """
 function solve!(x, solver::MultigridSolver, b; kwargs...)
     Nx, Ny, Nz = size(b)
 
     solver.b_array .= reshape(interior(b), Nx * Ny * Nz)
     solver.x_array .= reshape(interior(x), Nx * Ny * Nz)
 
-    solt = init(solver.amg_algorithm, solver.matrix, solver.b_array)
+    if architecture(solver) == CPU()
+        solt = init(solver.amg_algorithm, solver.matrix, solver.b_array)
+
+        _solve!(solver.x_array, solt.ml, solt.b, maxiter=solver.maxiter, abstol = solver.abstol, reltol=solver.reltol, kwargs...)
+    else 
+        s = solver.amgx_solver_struct
+        AMGX.upload!(s.device_b, solver.b_array)
+        AMGX.upload!(s.device_x, solver.x_array)
+        AMGX.setup!(s.amgx_solver, s.device_matrix)
+        AMGX.solve!(s.device_x, s.amgx_solver, s.device_b)
+        solver.x_array = CuArray(Vector(s.device_x)) #FIXME
+    end
 
-    _solve!(solver.x_array, solt.ml, solt.b, maxiter=solver.maxiter, abstol = solver.abstol, reltol=solver.reltol, kwargs...)
-    
     interior(x) .= reshape(solver.x_array, Nx, Ny, Nz)
 end
 
+function finalize_solver(solver::MultigridSolver)
+    s = solver.amgx_solver_struct
+    if s != nothing
+        close(s.device_matrix); close(s.device_x); close(s.device_b); close(s.amgx_solver); close(s.resources); close(s.config); 
+        AMGX.finalize_plugins(); AMGX.finalize()
+    end
+end
```

Comment:
few comments here

1. `finalize_solver` seems to modify its argument so perhaps rename to `finalize_solver!`?
2. instead of having an `if` statement use multiple dispatch?
  ```Julia
  julia> f(x, y) = x+y
  f (generic function with 1 method)

  julia> f(x, ::Nothing) = 2x
  f (generic function with 2 methods)

  julia> f(2, 3)
  5

  julia> f(2, nothing)
  4
  ```

So perhaps something like:

```Julia
finalize_solver!(solver::MultigridSolver) = finalize_solver!(solver.amgx_solver_struct)
finalize_solver!(::Nothing) = nothing
function finalize_solver!(s::SomeOtherTypeOfTheAMGXMultigridSolver)
    close(s.device_matrix)
    close(s.device_x)
    close(s.device_b)
    close(s.amgx_solver)
    close(s.resources)
    close(s.config)
    AMGX.finalize_plugins()
    AMGX.finalize()
 end
```

Reply:
I think this sort of philosophy can be used in other parts of this PR?